--- hb.cc	2022-07-04 20:31:35.190512039 +0900
+++ hb2.cc	2022-07-04 20:29:39.365472652 +0900
@@ -51,16 +51,10 @@
     {
       for (auto &[key, val] : dp[i])
         {
-          if (dp[i + 1].count(key) == 0)
-            {
-              dp[i + 1][key] = inf;
-            }
-          if (dp[i + 1].count(key + v[i]) == 0)
-            {
-              dp[i + 1][key + v[i]] = inf;
-            }
-          chmin(dp[i + 1][key], dp[i][key]);
-          chmin(dp[i + 1][key + v[i]], dp[i][key] + w[i]);
+          auto [it_key, _dummy1] = dp[i + 1].insert({key, inf});
+          auto [it_key_vi, _dummy2] = dp[i + 1].insert({key + v[i], inf});
+          chmin(it_key->second, val);
+          chmin(it_key_vi->second, val + w[i]);
         }
     }
   ll ans = 0;
